Provide ActivityScenario support

Time ago, Espresso core deprecated ActivityTestRule in favor or the ActivityScenario feature which is way more flexible. We should provide proper support for this new API ensuring at least one of our composers use this API. Documentation should also be updated.

Right now running in `scenario.onActivity` causes a crash:
"java.lang.RuntimeException: java.lang.RuntimeException: This method can not be called from the main application thread"

This makes some sense, because the `onActivity` callback runs on the main thread. Through the stacktrace it's clear where it goes wrong: The default impl of `runOnUi` calls `Intrumentation.runOnMainSync`, which fails if you're already on the main thread. To work around this crash, I can override `runOnUI`:
```kotlin
override fun runOnUi(block: () -> Unit) {
    if (Thread.currentThread() == Looper.getMainLooper().thread)
        block()
    else
        super.runOnUi(block)
}
```

Then another error appears. Same message, this time pointing to `waitForAnimationsToFinish`, called by `disableFlakyComponentsAndWaitForIdle`. Again, I can override this and skip the "waitForIdle" bits from the default implementation (I also have to skip some private method calls but hopefully those are unimportant 😬):
```kotlin
override fun disableFlakyComponentsAndWaitForIdle(view: View) {
    if (Thread.currentThread() == Looper.getMainLooper().thread)
        prepareUIForScreenshot()
    else
        super.disableFlakyComponentsAndWaitForIdle(view)
}
```

Now the test runs to completion!

But record mode is broken.
```
* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':demo:debugDownloadScreenshots'.
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:205)
	at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:203)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:184)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
	at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
	at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: java.lang.RuntimeException: Nonzero exit value: 1
	at scala.sys.package$.error(package.scala:27)
	at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.slurp(ProcessBuilderImpl.scala:134)
	at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$bang$bang(ProcessBuilderImpl.scala:105)
	at com.karumi.shot.android.Adb.executeAdbCommandWithResult(Adb.scala:45)
	at com.karumi.shot.android.Adb.pullScreenshots(Adb.scala:35)
	at com.karumi.shot.Shot.$anonfun$pullScreenshots$1(Shot.scala:147)
	at com.karumi.shot.Shot.$anonfun$pullScreenshots$1$adapted(Shot.scala:142)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at com.karumi.shot.Shot.pullScreenshots(Shot.scala:142)
	at com.karumi.shot.Shot.downloadScreenshots(Shot.scala:37)
	at com.karumi.shot.tasks.DownloadScreenshotsTask.downloadScreenshots(Tasks.scala:101)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:49)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:42)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:727)
	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:694)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.run(ExecuteActionsTaskExecuter.java:568)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:553)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:536)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$300(ExecuteActionsTaskExecuter.java:109)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.executeWithPreviousOutputFiles(ExecuteActionsTaskExecuter.java:276)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:265)
	at org.gradle.internal.execution.steps.ExecuteStep.lambda$execute$1(ExecuteStep.java:33)
	at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:33)
	at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:26)
	at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:67)
	at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:36)
	at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:49)
	at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:34)
	at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:43)
	at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:73)
	at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:54)
	at org.gradle.internal.execution.steps.CatchExceptionStep.execute(CatchExceptionStep.java:34)
	at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:44)
	at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:54)
	at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:38)
	at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:49)
	at org.gradle.internal.execution.steps.CacheStep.executeWithoutCache(CacheStep.java:159)
	at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:72)
	at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:43)
	at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:44)
	at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:33)
	at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
	at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:92)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:85)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:55)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:39)
	at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:76)
	at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
	at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:94)
	at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:49)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:79)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:53)
	at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:74)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:78)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:78)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:34)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:39)
	at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:40)
	at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:28)
	at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:192)
	... 30 more
```

This failure may be related to one or more of a few things:
* I didn't call those private methods when overriding `disableFlakyComponentsAndWaitForIdle`.
* The test I'm running is the only one in the suite.

Any ideas for what to look at next?

--- a/README.md
+++ b/README.md
@@ -203,6 +203,23 @@ dependencies {
 
 Keep in mind you'll need to use a compatible version or Shot won't work as expected.
 
+## ActivityScenario support
+
+``ActivityTestRule`` has been deprecated and now the usage of ActivityScenario is recommended. However, Shot needs to be executed from the instrumentation thread to be able to extract all the test metadata needed to record and verify screenshots. That's why we've created an ``ActivityScenario`` extension method named ``waitForActivity``. This extension is needed get the activity instance from the instrumentation thread instead of running Shot from the app target thread. Using this extension you can write tests like this:
+
+```kotlin
+class MainActivityTest: ScreenshotTest {
+    @Test
+    fun rendersTheDefaultActivityState() {
+        val activity = ActivityScenario.launch(MainActivity::class.java)
+
+        compareScreenshot(activity)
+    }
+}
+```
+
+I hope we can find a better solution for this issue in the future.
+
 ## Recording tests
 
 You can record your screenshot tests executing this command:

--- a/shot-android/src/main/java/com/karumi/shot/ActivityScenarioUtils.kt
+++ b/shot-android/src/main/java/com/karumi/shot/ActivityScenarioUtils.kt
@@ -1,0 +1,23 @@
+package com.karumi.shot
+
+import android.app.Activity
+import androidx.test.core.app.ActivityScenario
+import java.lang.IllegalStateException
+
+/* ActivityTestRule has been deprecated and now the usage of ActivityScenario is recommended.
+ * However, Shot needs to be executed from the instrumentation thread to be able to extract
+ * all the test metadata needed to record and verify screenshots. That's why we've created this
+ * extension to be able to get the activity instance from the instrumentation thread instead
+ * of running Shot from the app target thread. I hope we can find a better solution in the future.
+ */
+fun <A : Activity> ActivityScenario<A>.waitForActivity(): A {
+    var activity: A? = null
+    onActivity {
+        activity = it
+    }
+    return if (activity != null) {
+        activity!!
+    } else {
+        throw IllegalStateException("The activity scenario could not be initialized.")
+    }
+}

--- a/shot-android/src/main/java/com/karumi/shot/ScreenshotTest.kt
+++ b/shot-android/src/main/java/com/karumi/shot/ScreenshotTest.kt
@@ -4,6 +4,7 @@ import android.app.Activity
 import android.app.Dialog
 import android.content.Context
 import android.os.Build
+import android.os.Looper
 import android.util.DisplayMetrics
 import android.util.Log
 import android.view.View
@@ -100,9 +101,9 @@ interface ScreenshotTest {
         val width = widthInPx ?: metrics.widthPixels
         runOnUi {
             ViewHelpers.setupView(view)
-                .setExactHeightPx(height)
-                .setExactWidthPx(width)
-                .layout()
+                    .setExactHeightPx(height)
+                    .setExactWidthPx(width)
+                    .layout()
         }
         takeViewSnapshot(name, view)
     }
@@ -128,7 +129,9 @@ interface ScreenshotTest {
             disableAnimatedComponents(view)
             hideIgnoredViews(view)
         }
-        waitForAnimationsToFinish()
+        if (notInAppMainThread()) {
+            waitForAnimationsToFinish()
+        }
     }
 
     private fun hideIgnoredViews(view: View) = runOnUi {
@@ -143,18 +146,24 @@ interface ScreenshotTest {
     }
 
     fun runOnUi(block: () -> Unit) {
-        getInstrumentation().runOnMainSync { block() }
+        if (notInAppMainThread()) {
+            getInstrumentation().runOnMainSync { block() }
+        } else {
+            block()
+        }
     }
+
+    private fun notInAppMainThread() = Looper.myLooper() != Looper.getMainLooper()
 
     private fun takeViewSnapshot(name: String?, view: View) {
         val testName = name ?: TestNameDetector.getTestName()
         val snapshotName = "${TestNameDetector.getTestClass()}_$testName"
         try {
             Screenshot
-                .snap(view)
-                .setIncludeAccessibilityInfo(false)
-                .setName(snapshotName)
-                .record()
+                    .snap(view)
+                    .setIncludeAccessibilityInfo(false)
+                    .setName(snapshotName)
+                    .record()
         } catch (t: Throwable) {
             Log.e("Shot", "Exception captured while taking screenshot for snapshot with name $snapshotName", t)
         }
@@ -165,10 +174,10 @@ interface ScreenshotTest {
         val snapshotName = "${TestNameDetector.getTestClass()}_$testName"
         try {
             Screenshot
-                .snapActivity(activity)
-                .setIncludeAccessibilityInfo(false)
-                .setName(snapshotName)
-                .record()
+                    .snapActivity(activity)
+                    .setIncludeAccessibilityInfo(false)
+                    .setName(snapshotName)
+                    .record()
         } catch (t: Throwable) {
             Log.e("Shot", "Exception captured while taking screenshot for snapshot with name $snapshotName", t)
         }

--- a/shot-consumer-flavors/app/build.gradle
+++ b/shot-consumer-flavors/app/build.gradle
@@ -78,12 +78,12 @@ dependencies {
 
     testImplementation "junit:junit:4.12"
 
-    androidTestImplementation "androidx.test:runner:1.2.0"
-    androidTestImplementation "androidx.test.espresso:espresso-core:3.2.0"
-    androidTestImplementation "androidx.test.espresso:espresso-intents:3.2.0"
-    androidTestImplementation "androidx.test.espresso:espresso-contrib:3.2.0"
-    androidTestImplementation "androidx.test:rules:1.2.0"
-    androidTestImplementation "androidx.test.ext:junit:1.1.1"
+    androidTestImplementation "androidx.test:runner:1.3.0"
+    androidTestImplementation "androidx.test.espresso:espresso-core:3.3.0"
+    androidTestImplementation "androidx.test.espresso:espresso-intents:3.3.0"
+    androidTestImplementation "androidx.test.espresso:espresso-contrib:3.3.0"
+    androidTestImplementation "androidx.test:rules:1.3.0"
+    androidTestImplementation "androidx.test.ext:junit:1.1.2"
     androidTestImplementation "asm:asm:3.3.1"
     androidTestImplementation "org.jetbrains.kotlinx:kotlinx-coroutines-test:1.1.1"
     androidTestImplementation "com.github.tmurakami:dexopener:2.0.0"

--- a/shot-consumer-flavors/app/src/androidTest/java/com/karumi/ui/view/AcceptanceTest.kt
+++ b/shot-consumer-flavors/app/src/androidTest/java/com/karumi/ui/view/AcceptanceTest.kt
@@ -3,25 +3,25 @@ package com.karumi.ui.view
 import android.app.Activity
 import android.content.Intent
 import android.os.Bundle
-import androidx.test.espresso.intent.rule.IntentsTestRule
+import androidx.test.core.app.ActivityScenario
+import androidx.test.core.app.ApplicationProvider
+import androidx.test.espresso.intent.Intents
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import androidx.test.filters.LargeTest
 import androidx.test.platform.app.InstrumentationRegistry
 import com.github.salomonbrys.kodein.Kodein
 import com.karumi.asApp
 import com.karumi.shot.ScreenshotTest
+import org.junit.After
 import org.junit.Before
-import org.junit.Rule
 import org.junit.runner.RunWith
 import org.mockito.MockitoAnnotations
+import com.karumi.shot.waitForActivity
 
 @LargeTest
 @RunWith(AndroidJUnit4::class)
-abstract class AcceptanceTest<T : Activity>(clazz: Class<T>) : ScreenshotTest {
-
-    @Rule
-    @JvmField
-    val testRule: IntentsTestRule<T> = IntentsTestRule(clazz, true, false)
+abstract class AcceptanceTest<T : Activity>(private val clazz: Class<T>) : ScreenshotTest {
+    private var scenario: ActivityScenario<T>? = null
 
     @Before
     fun setup() {
@@ -29,12 +29,20 @@ abstract class AcceptanceTest<T : Activity>(clazz: Class<T>) : ScreenshotTest {
         val app = InstrumentationRegistry.getInstrumentation().targetContext.asApp()
         app.resetInjection()
         app.overrideModule = testDependencies
+        Intents.init()
+    }
+
+    @After
+    fun tearDown() {
+        scenario?.close()
+        Intents.release()
     }
 
     fun startActivity(args: Bundle = Bundle()): T {
-        val intent = Intent()
+        val intent = Intent(ApplicationProvider.getApplicationContext(), clazz)
         intent.putExtras(args)
-        return testRule.launchActivity(intent)
+        scenario = ActivityScenario.launch(intent)
+        return scenario!!.waitForActivity()
     }
 
     abstract val testDependencies: Kodein.Module

